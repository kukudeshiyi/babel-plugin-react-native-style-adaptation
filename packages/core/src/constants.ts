export const PLUGIN_NAME = 'babel-plugin-react-native-style-adaptation';
