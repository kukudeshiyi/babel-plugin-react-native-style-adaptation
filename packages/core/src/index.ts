import { PluginObj } from '@babel/core';
import { NodePath } from '@babel/core';
import * as t from '@babel/types';
export { distance, font } from './properties';

export const PLUGIN_NAME = 'babel-plugin-react-native-style-sheet';
const validateStatus = Symbol(PLUGIN_NAME);

interface Config {
  properties: string[];
  module: string;
  source: string;
}

interface State {
  [validateStatus]: boolean;
  opts: { configs: Config[] };
}

export default (): PluginObj<State> => {
  return {
    name: PLUGIN_NAME,
    pre(state) {
      state[validateStatus] = validate(state.opts?.configs || []);
    },
    visitor: {
      ObjectProperty(path, state) {
        normalTransform(path, state);
      },
    },
  };
};

function normalTransform(path: NodePath<t.ObjectProperty>, state: State) {
  if (!state[validateStatus]) {
    return;
  }

  const configs = state.opts.configs;
  const propertyName = path.node.key.name || '';
  const config = configs.find((config) => config.properties.includes(propertyName));
  if (!config) {
    return;
  }

  const propertyValue = path.node.value;
  if (!t.isStringLiteral(propertyValue) && !t.isNumericLiteral(propertyValue)) {
    return;
  }

  const module = config.module;
  const source = config.source;

  const node = addAdaptation(propertyValue.value, module);
  path.node.value = node;
  injectImportExpression(path, module, source);
}

function injectImportExpression(
  currentPath: NodePath<t.ObjectProperty>,
  module: Config['module'],
  source: Config['source'],
) {
  const programPath = currentPath.findParent((parentNode) => t.isProgram(parentNode));
  const find = (callback: (path: NodePath<t.Node>) => boolean) =>
    programPath.get('body').find(callback);
  const firstImportNodePath = find((nodePath) => t.isImportDeclaration(nodePath.node));

  if (!firstImportNodePath && isProgramNode(programPath.node)) {
    return programPath.node.body.unshift(createImportDeclarationNode(module, source));
  }

  const matchedLibraryPath = find(
    (nodePath) => t.isImportDeclaration(nodePath.node) && nodePath.node.source.value === source,
  );

  if (!matchedLibraryPath && firstImportNodePath) {
    return firstImportNodePath.replaceWithMultiple([
      firstImportNodePath.node,
      createImportDeclarationNode(module, source),
    ]);
  }

  // fix type error
  if (matchedLibraryPath && isImportDeclarationNode(matchedLibraryPath.node)) {
    const importNode = matchedLibraryPath.node;
    if (!importNode.specifiers) {
      // TODO: error
      return;
    }

    const isModuleExist = importNode.specifiers.find((v) => v.local.name === module);
    if (isModuleExist) {
      return;
    }
    importNode.specifiers.push(t.importSpecifier(t.identifier(module), t.identifier(module)));
  }
}

function validate(configs: Config[]) {
  if (!Array.isArray(configs) || configs.length <= 0) {
    //TODO: error
    console.log('error', 1);
    return false;
  }
  for (const config of configs) {
    const { properties, module, source } = config;
    if (!Array.isArray(properties) || properties.length <= 0) {
      //TODO: error
      console.log('error', 2);
      return false;
    }
    if (typeof module !== 'string') {
      //TODO: error
      console.log('error', 3);
      return false;
    }
    if (typeof source !== 'string') {
      //TODO: error
      console.log('error', 4);
      return false;
    }
  }
  return true;
}

function addAdaptation(value: string | number, module: Config['module']) {
  return t.callExpression(t.identifier(module), [
    isString(value) ? t.stringLiteral(value) : t.numericLiteral(value),
  ]);
}

function createImportDeclarationNode(module: Config['module'], source: Config['source']) {
  return t.importDeclaration(
    [t.importSpecifier(t.identifier(module), t.identifier(module))],
    t.stringLiteral(source),
  );
}

function isString(value: any): value is string {
  return typeof value === 'string';
}

function isProgramNode(node: any): node is t.Program {
  return t.isProgram(node);
}

function isImportDeclarationNode(node: any): node is t.ImportDeclaration {
  return t.isImportDeclaration(node);
}
