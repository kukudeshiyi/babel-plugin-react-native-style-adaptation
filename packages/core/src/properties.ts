// export const allowProperties = [
//   'margin',
//   'padding',
//   'borderRadius',
//   'border',
//   'borderTop',
//   'borderLeft',
//   'borderBottom',
//   'borderRight',
//   'font',
//   'boxShadow',
//   'background',
//   'flex',
//   'flexFlow',
//   'fontFamily',
//   'fontVariant',
//   'shadowOffset',
//   'textShadow',
//   'textShadowOffset',
//   'transform',
// ];

// export const lengthProperties = [
//   'borderBottomEndRadius',
//   'borderBottomLeftRadius',
//   'borderBottomRightRadius',
//   'borderBottomStartRadius',
//   'borderBottomWidth',
//   'borderLeftWidth',
//   'borderRadius',
//   'borderRightWidth',
//   'borderTopEndRadius',
//   'borderTopLeftRadius',
//   'borderTopRightRadius',
//   'borderTopStartRadius',
//   'borderTopWidth',
//   'borderWidth',
//   'borderEndWidth',
//   'borderStartWidth',
//   'bottom',
//   'height',
//   'left',
//   'margin',
//   'marginBottom',
//   'marginEnd',
//   'marginHorizontal',
//   'marginLeft',
//   'marginRight',
//   'marginStart',
//   'marginTop',
//   'marginVertical',
//   'maxHeight',
//   'maxWidth',
//   'minHeight',
//   'minWidth',
//   'padding',
//   'paddingBottom',
//   'paddingEnd',
//   'paddingHorizontal',
//   'paddingLeft',
//   'paddingRight',
//   'paddingStart',
//   'paddingTop',
//   'paddingVertical',
//   'position',
//   'right',
//   'top',
//   'width',
//   'textShadowOffset',
//   'fontSize',
//   'lineHeight',
//   'textShadowRadius',
//   'includeFontPadding',
//   'letterSpacing',
//   'shadowRadius',
// ];

// export default [...allowProperties, ...lengthProperties];

export const distance = ['margin', 'padding'];
export const font = ['fontSize'];
