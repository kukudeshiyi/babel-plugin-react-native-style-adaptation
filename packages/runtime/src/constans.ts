import { Dimensions } from 'react-native';

export const DEFAULT_DESIGN_DRAFT_WIDTH = 375;
export const WINDOW_WIDTH = Dimensions.get('window').width;
