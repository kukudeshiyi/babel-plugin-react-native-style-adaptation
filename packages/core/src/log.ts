import chalk from 'chalk';
import { PLUGIN_NAME } from './constants';

export enum LOG_TYPE {
  INFO = 'blue',
  WARNING = 'yellow',
  ERROR = 'red',
}

function log(type: LOG_TYPE, message: string) {
  console.log(chalk[type](message));
}

export function logInfo(message: string) {
  const infoMessage = `[${PLUGIN_NAME} info]: ${message}`;
  log(LOG_TYPE.INFO, infoMessage);
}

export function logWarning(message: string) {
  const warningMessage = `[${PLUGIN_NAME} warning]: ${message}`;
  log(LOG_TYPE.WARNING, warningMessage);
}

export function logError(message: string) {
  const errorMessage = `[${PLUGIN_NAME} error]: ${message}`;
  log(LOG_TYPE.ERROR, errorMessage);
  process.exit(1);
}
