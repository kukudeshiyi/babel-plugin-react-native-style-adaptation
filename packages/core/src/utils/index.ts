import { NodePath } from '@babel/core';
import * as t from '@babel/types';

/**
 * 插入import
 * 判断 import 是否有同名
 * 如果没有则插入一条新的
 * 如果有，
 * 则判断是否有导出同名的 ident,
 * 如果有则跳过，
 * 如果没有则替换
 *
 * @param {*} bpath ast path
 * @param {*} source 依赖名
 * @param {*} ident 导出字段名
 */
export function injectImportExpression(bpath: NodePath, source: string, ident: string) {
  const programPath = bpath.findParent((parentNode) => t.isProgram(parentNode));
  const find = (callback: (path: NodePath<t.Node>) => boolean) =>
    programPath.get('body').find(callback);

  const firstImportNodePath = find((nodePath) => t.isImportDeclaration(nodePath.node));

  const generatorImportStrcut = (idents: Array<string>) =>
    t.importDeclaration(
      idents.map((curId: string) => t.importSpecifier(t.identifier(curId), t.identifier(curId))),
      t.stringLiteral(source),
    );

  if (!firstImportNodePath) {
    // program.get('body').unshiftContainer
    (programPath.node as t.Program).body.unshift(generatorImportStrcut([ident]));
    return;
  }

  const matchedLibraryPath = find(
    (nodePath) => t.isImportDeclaration(nodePath.node) && nodePath.node.source.value === source,
  );

  if (!matchedLibraryPath) {
    firstImportNodePath.replaceWithMultiple([
      firstImportNodePath.node,
      generatorImportStrcut([ident]),
    ]);
    return;
  }

  const importNode = matchedLibraryPath.node as t.ImportDeclaration;
  if (!importNode.specifiers) {
    console.log('依赖没有 specifier, 请检查', source);
    return;
  }

  const existIdentifier = importNode.specifiers.find((v) => v.local.name === ident);
  // 已经导入了，则跳过
  if (existIdentifier) return;
  importNode.specifiers.push(t.importSpecifier(t.identifier(ident), t.identifier(ident)));
}
