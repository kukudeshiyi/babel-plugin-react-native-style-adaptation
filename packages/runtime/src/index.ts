import { WINDOW_WIDTH, DEFAULT_DESIGN_DRAFT_WIDTH } from './constans';
import { InitParams } from './types';

export { WINDOW_WIDTH, DEFAULT_DESIGN_DRAFT_WIDTH } from './constans';
export { InitParams } from './types';

export let designDraftWidth: InitParams['designDraftWidth'] = DEFAULT_DESIGN_DRAFT_WIDTH;

export function scaleSize(size: number) {
  return size * (WINDOW_WIDTH / designDraftWidth);
}

export default function init(params: InitParams) {
  const { designDraftWidth: customDesignDraftWidth = DEFAULT_DESIGN_DRAFT_WIDTH } = params;
  if (typeof customDesignDraftWidth === 'number') {
    designDraftWidth = customDesignDraftWidth;
  }
}
