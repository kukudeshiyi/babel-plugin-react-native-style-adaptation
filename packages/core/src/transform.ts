import { NodePath } from '@babel/core';
import * as t from '@babel/types';
import { getStylesForProperty } from 'css-to-react-native';
import StyleTable from './style-table';
import { injectImportExpression } from './utils';
import { isEqual } from 'lodash';

const supportUnit = ['rpx', 'pt', 'px', 'vh', 'vw'];

export function normalTransform(path: NodePath<t.ObjectProperty>) {
  if (!StyleTable.includes(path.node.key?.name)) {
    return;
  }

  // 只会转换 string 和 number, 其他表达式不转换
  if (!t.isStringLiteral(path.node.value) && !t.isNumericLiteral(path.node.value)) return;

  const styleSheet = getStylesForProperty(path.node.key?.name, String(path.node.value.value));

  // 无需转换
  if (
    isEqual(styleSheet, { [path.node.key?.name]: path.node.value.value }) &&
    !isSuportUnit(path.node.value.value)
  ) {
    return;
  }

  const importLibrarys = new Set();
  const nodes: t.ObjectProperty[] = [];

  // 转换成 babel ast
  Object.keys(styleSheet).forEach((key: string) => {
    const value = styleSheet[key];

    if (isNumber(value)) {
      nodes.push(t.objectProperty(t.identifier(key), t.numericLiteral(Number(value))));
      return;
    }

    if (isSuportUnit(value)) {
      const unit = getUnit(value);
      const exportVariable = `__$${unit}`;
      importLibrarys.add(exportVariable);

      nodes.push(
        t.objectProperty(
          t.identifier(key),
          t.callExpression(t.identifier(exportVariable), [
            t.numericLiteral(Number(value.substr(0, value.length - unit.length))),
          ]),
        ),
      );
      return;
    }

    nodes.push(t.objectProperty(t.identifier(key), t.stringLiteral(value)));
  });

  importLibrarys.forEach((vari: any) => {
    injectImportExpression(path as any, 'react-native-style-runtime', vari as string);
  });

  path.replaceWithMultiple(nodes);
}

function isNumber(value: any): boolean {
  return value * 1 === value * 1;
}

function isSuportUnit(value: any): boolean {
  return supportUnit.some((unit) => String(value).endsWith(unit));
}

function getUnit(value: string): string {
  return supportUnit.filter((unit) => value.endsWith(unit))[0];
}
