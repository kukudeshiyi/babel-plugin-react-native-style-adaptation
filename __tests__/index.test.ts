import test from 'babel-plugin-tester';
import plugin, { PLUGIN_NAME } from '../packages/core/src/index';
import { distance, font } from '../packages/core/src/properties';

test({
  plugin,
  pluginOptions: {
    configs: [
      {
        properties: distance,
        module: 'scaleSize',
        source: 'scaleSizeLib',
      },
      {
        properties: font,
        module: 'scaleFont',
        source: 'scaleFontLib',
      },
    ],
  },
  pluginName: PLUGIN_NAME,
  tests: [
    {
      code: `
          const styles = StyleSheet.create({
            margin:10,
            padding:20
          })
        `,
      snapshot: true,
    },
  ],
});
