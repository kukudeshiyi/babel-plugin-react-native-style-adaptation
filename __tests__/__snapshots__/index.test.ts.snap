// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-react-native-style-sheet 1. babel-plugin-react-native-style-sheet: 1. babel-plugin-react-native-style-sheet 1`] = `

const styles = StyleSheet.create({
  margin:10,
  padding:20
})

      ↓ ↓ ↓ ↓ ↓ ↓

import { scaleSize } from 'scaleSizeLib';
const styles = StyleSheet.create({
  margin: scaleSize(10),
  padding: scaleSize(20),
});


`;
